toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,293 @@
+  "secret.txt": "my-secret"
+}
+```
+
+The above `secret` would mount files named `application.yml` and `secret.txt` at the path configured in the binding.
+
+### Creating a `secret`
+
+To create a `secret`, users will use a new CLI command: `cf create-secret`.
+
+```console
+cf create-secret SECRET_NAME -t opaque -p <CREDENTIALS_INLINE>
+```
+
+```console
+cf create-secret SECRET_NAME -t opaque -f <FILE_PATH_TO_CREDENTIALS_FILE>
+```
+
+* `-t` the secret `type`, defaults to `opaque`
+* `-p` reads inline/literal credentials
+* `-f` reads credentials from a file
+
+This command will ultimately create a `POST` request to `/v3/secrets`. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets" \
+  -X POST \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "opaque",
+    "name": "my-app-secrets",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "my-secret"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    },
+    "relationships": {
+      "space": {
+        "data": {
+          "guid": "7304bc3c-7010-11ea-8840-48bf6bec2d78"
+        }
+      }
+    }
+  }'
+```
+
+### Updating a `secret`
+Users will be able to update a `secret` using similar commands:
+
+```console
+cf update-secret SECRET_NAME SECRET_TYPE -p <CREDENTIALS_INLINE>
+```
+
+```console
+cf update-secret SECRET_NAME SECRET_TYPE -f <FILE_PATH_TO_CREDENTIALS_FILE>
+```
+
+Initially apps will need to be restarted to receive `secret` updates, but eventually we may be able to support updates that don't require a restart.
+The "Update secret" endpoint in the CF API will take a `PATCH` request. Example `curl` invocation:
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]" \
+  -X PATCH \
+  -H "Authorization: bearer [token]" \
+  -H "Content-type: application/json" \
+  -d '{
+    "type": "opaque",
+    "value": {
+      "application.yml" : "---\napplication:\n ...",
+      "secret.txt": "updated-secret-value"
+    }
+    "metadata": {
+      "annotations": {},
+      "labels": {}
+    }
+  }'
+```
+
+### Reading a `secret` and Its Value
+Users that have permission to view environment variables for an app will be able to view `secret` values within a `space`. We will introduce a separate endpoint for retrieving these (similar to how app environment variables and service instance credentials are retrieved): `GET /v3/secrets/:guid/value`.
+
+```console
+curl "https://api.example.org/v3/secrets/[guid]/value" \
+  -X GET \
+  -H "Authorization: bearer [token]"
+  
+"value" : {
+  "some-json-key" : "some-json-value"
+}
+```
+
+To support this, Cloud Controller will need to have both read and write permissions for the `secret`'s corresponding credential in CredHub.
+
+### Creating a `secret_binding`
+To create `secret_bindings`, users will use a new CLI command: `cf bind-secret`. This will create a `POST` request to `/v3/secret_bindings`.
```

Comment:
Is this going to force a reload of the application to ensure that secrets will be available or the user will have to do it afterward manually?

Reply:
I think for a new binding developers will need to `cf restart` their app the same way they do to make env vars and service binding changes take effect. I believe this is the same behavior as Kubernetes when you add new volume mounts to a Pod spec. For updates of existing bindings I think the same is true, at least to start, but we could explore secret changes not requiring a restart in the future.